File: app/src/main/java/com/kaciula/archiman/presentation/util/BaseController.java
package com.kaciula.archiman.presentation.util;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.ActionBar;
import android.view.View;
import com.bluelinelabs.conductor.Controller;
import com.kaciula.archiman.presentation.screen.main.ActionBarProvider;
import com.kaciula.archiman.presentation.screen.main.DialogShowman;

public abstract class BaseController extends RefWatchingController {

  protected BaseController() {
  }

  protected BaseController(Bundle args) {
    super(args);
  }

  protected ActionBar getActionBar() {
    ActionBarProvider actionBarProvider = ((ActionBarProvider) getActivity());
    return actionBarProvider != null ? actionBarProvider.getSupportActionBar() : null;
  }

  @Override
  protected void onAttach(@NonNull View view) {
    setTitle();
    super.onAttach(view);
  }

  protected void setTitle() {
    Controller parentController = getParentController();
    while (parentController != null) {
      if (parentController instanceof BaseController
          && ((BaseController) parentController).getTitle() != null) {
        return;
      }
      parentController = parentController.getParentController();
    }

    String title = getTitle();
    ActionBar actionBar = getActionBar();
    if (title != null && actionBar != null) {
      actionBar.setTitle(title);
    }
  }

  protected String getTitle() {
    return null;
  }

  protected DialogShowman getDialogShowman() {
    return (DialogShowman) getActivity();
  }
}

File: app/src/main/java/com/kaciula/archiman/presentation/util/BaseController.kt
package com.kaciula.archiman.presentation.util

import android.os.Bundle
import android.support.v7.app.ActionBar
import android.view.View
import com.kaciula.archiman.presentation.screen.main.ActionBarProvider
import com.kaciula.archiman.presentation.screen.main.DialogShowman
import com.kaciula.archiman.presentation.screen.main.Navigator
import timber.log.Timber

abstract class BaseController : ButterKnifeController {

    protected constructor()

    protected constructor(args: Bundle) : super(args)

    protected fun getActionBar(): ActionBar? {
        val actionBarProvider = activity as ActionBarProvider?
        return actionBarProvider?.getSupportActionBar()
    }

    override fun onAttach(view: View) {
        Timber.i("ATTACH CONTROLLER -> ${javaClass.simpleName}")
        setTitle()
        super.onAttach(view)
    }

    override fun onDetach(view: View) {
        Timber.i("DETACH CONTROLLER -> ${javaClass.simpleName}")
        super.onDetach(view)
    }

    protected fun setTitle() {
        var parentController = parentController
        while (parentController != null) {
            if (parentController is BaseController && parentController.getTitle() != null) {
                return
            }
            parentController = parentController.parentController
        }

        if (getTitle() != null) {
            getActionBar()?.title = getTitle()
        }
    }

    protected fun getTitle(): String? = null

    protected fun getDialogShowman(): DialogShowman = activity as DialogShowman

    protected fun getNavigator(): Navigator = activity as Navigator
}
